client/core/core.go

```diff
@@ -4572,7 +4573,30 @@ func handleMatchProofMsg(c *Core, dc *dexConnection, msg *msgjson.Message) error
 			note.MarketID)
 	}
 
-	return book.ValidateMatchProof(note)
+	err = book.ValidateMatchProof(note)
+	if err != nil {
+		return fmt.Errorf("match proof validation failed: %w", err)
+	}
+
+	// Validate server match checksum for client orders (trades) in this epoch.
+	for _, trade := range dc.trackedTrades() {
+		// TODO - think about len(trade.csum) > 0 fits with the rest of these conditions, and at least document the reason for having it.
```

Comment:
Is it possible for `processPreimageRequest` to have not happened yet?

I don't see a check for `nil` in `processPreimageRequest`, unless I am missing a check somewhere else. Maybe server can send nil in the `PreimageRequest` request?

Reply:
Yes, I've mentioned that in my comment above (section 3), we should probably check against that,

https://github.com/decred/dcrdex/pull/1077#discussion_r628744407

Reply:
Ah, somehow I missed that long conversation. Well, I don't know the best answer, but thank you for your time on this. imo coming along nicely!